ext/cl_khr_command_buffer.asciidoc

```diff
@@ -492,6 +492,26 @@ with the command-queue used on command-buffer creation. Where a _compatible_
 command-queue is defined as a command-queue with identical properties targeting
 the same device and in the same OpenCL context.
 
+While constructing a command-buffer it is valid for the user to interleave calls
+to the same queue which create commands, such as {clCommandNDRangeKernelKHR}, with
+queue submission calls, such as {clEnqueueNDRangeKernel} or
+{clEnqueueCommandBufferKHR}. That is, there is no effect on queue state from
+recording commands. The purpose of the queue parameter is to define the device
+and properties of the command, which are constant queries on the queue object.
+
+A command-buffer object should increment the reference count of attached OpenCL
+objects such as queues, buffers, images, and kernels referenced in commands
+recorded to the command-buffer. This enables correct behavior of the
+command-buffer when its attached objects have been released. On destruction of
+the command-buffer it should decrement these reference counts, allowing the
+attached objects to be freed if appropriate.
+
+[NOTE]
+====
+A command-buffer object does not update the reference count of objects set as
+arguments on kernels recorded into the command-buffer. This is consistent with
+the reference counting behavior of {clSetKernelArg}.
```

Comment:
It feels odd that arguments to kernels are not refcounted but other objects used in commands are. We should be consistent in our approach.

Reply:
> My perspective on this was that we are updating the ref counts on the OpenCL objects that are passed to the command creation entry-points, and for clCommandNDRangeKernelKHR that is a cl_kernel, so we're updating the ref count of the cl_kernel object.

This was the rationale I was thinking of also, and I like the symmetry with clSetKernelArg.

Note, now that we have memory object destructor callbacks it's probably safer to retain references to memory objects then it was back in the old OpenCL 1.0 days.  If we decide to "fix" this though, I think we should do so consistently, in some future version of OpenCL.

(Selfishly, it would also be really hard or impossible to retain kernel arguments in my command buffer emulation layer, though if the kernel held reference counts to its kernel arguments then this should "just work".)